optimize/qa/schema-integrity-tests/src/test/java/io/camunda/optimize/test/upgrade/AbstractDatabaseSchemaIT.java

```diff
@@ -0,0 +1,166 @@
+    upgradeOutputWriter = new FileWriter(getOptimizeUpdateLogPath());
+  }
+
+  @AfterEach
+  public void logFileFlush() throws IOException {
+    oldOptimizeOutputWriter.flush();
+    oldOptimizeOutputWriter.close();
+    newOptimizeOutputWriter.flush();
+    newOptimizeOutputWriter.close();
+    upgradeOutputWriter.flush();
+    upgradeOutputWriter.close();
+  }
+
+  @Test
+  public void upgradeAndVerifySchemaIntegrityTest() throws Exception {
+    initializeClientAndCleanDatabase();
+    final OptimizeWrapper oldOptimize =
+        new OptimizeWrapper(getDatabaseType(), previousVersion, buildDirectory, oldDatabasePort);
+    final OptimizeWrapper newOptimize =
+        new OptimizeWrapper(getDatabaseType(), currentVersion, buildDirectory, newDatabasePort);
+
+    try {
+      // start new optimize to obtain expected schema and settings
+      newOptimize.start(getNewOptimizeOutputLogPath());
+      newDatabaseSchemaClient.refreshAll();
+      saveNewOptimizeDatabaseStatus();
+      newOptimize.stop();
+      newDatabaseSchemaClient.cleanIndicesAndTemplates();
+
+      // start old optimize to prepare for upgrade
+      oldOptimize.start(getOldOptimizeOutputLogPath());
+      oldDatabaseSchemaClient.refreshAll();
+      oldOptimize.stop();
+
+      // perform snapshot and restore on new es
+      oldDatabaseSchemaClient.createSnapshotRepository();
+      oldDatabaseSchemaClient.createSnapshotOfOptimizeIndices();
+      oldDatabaseSchemaClient.deleteSnapshotRepository();
+
+      newDatabaseSchemaClient.createSnapshotRepository();
+      newDatabaseSchemaClient.restoreSnapshot();
+      newDatabaseSchemaClient.deleteSnapshot();
+
+      // start a new async snapshot operation to ensure the upgrade is resilient to concurrently
+      // running snapshots
+      newDatabaseSchemaClient.createAsyncSnapshot();
+
+      // run the upgrade
+      newOptimize.startUpgrade(getOptimizeUpdateLogPath());
+      newOptimize.waitForUpgradeToFinish(360);
+
+      // stop/delete async snapshot operation as upgrade completed already
+      newDatabaseSchemaClient.deleteAsyncSnapshot();
+
+      // start new optimize
+      newOptimize.start(getNewOptimizeOutputLogPath());
+      newOptimize.stop();
+
+      log.info("Asserting expected index metadata...");
+      assertMigratedDatabaseIndicesMatchExpected();
+      assertMigratedDatabaseAliasesMatchExpected();
+      assertMigratedDatabaseTemplatesMatchExpected();
+      log.info("Finished asserting expected index metadata!");
+    } finally {
+      oldDatabaseSchemaClient.close();
+      newDatabaseSchemaClient.close();
+      oldOptimize.stop();
+      newOptimize.stop();
+    }
+  }
+
+  protected Integer getOldDatabasePort() {
+    return oldDatabasePort;
+  }
+
+  protected Integer getNewDatabasePort() {
+    return newDatabasePort;
+  }
+
+  protected String getBuildDirectory() {
+    return buildDirectory;
+  }
+
+  protected <T> void assertMapContentEqualityFieldByField(
```

Comment:
why does this exist in the abstract class?

Reply:
I put it there in case we want to use it for ElasticSearch as well, the method is generic and will also work there